lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -52,6 +95,246 @@ struct ExprVisitor {
+
+    case BinaryOperator::BinaryAnd:
+      return createBinary<moore::AndOp>(lhs, rhs);
+    case BinaryOperator::BinaryOr:
+      return createBinary<moore::OrOp>(lhs, rhs);
+    case BinaryOperator::BinaryXor:
+      return createBinary<moore::XorOp>(lhs, rhs);
+    case BinaryOperator::BinaryXnor: {
+      auto result = createBinary<moore::XorOp>(lhs, rhs);
+      if (!result)
+        return {};
+      return builder.create<moore::NotOp>(loc, result);
+    }
+
+    case BinaryOperator::Equality:
+      return createBinary<moore::EqOp>(lhs, rhs);
+    case BinaryOperator::Inequality:
+      return createBinary<moore::NeOp>(lhs, rhs);
+    case BinaryOperator::CaseEquality:
+      return createBinary<moore::CaseEqOp>(lhs, rhs);
+    case BinaryOperator::CaseInequality:
+      return createBinary<moore::CaseNeOp>(lhs, rhs);
+    case BinaryOperator::WildcardEquality:
+      return createBinary<moore::WildcardEqOp>(lhs, rhs);
+    case BinaryOperator::WildcardInequality:
+      return createBinary<moore::WildcardNeOp>(lhs, rhs);
+
+    case BinaryOperator::GreaterThanEqual:
+      return createBinary<moore::GeOp>(lhs, rhs);
+    case BinaryOperator::GreaterThan:
+      return createBinary<moore::GtOp>(lhs, rhs);
+    case BinaryOperator::LessThanEqual:
+      return createBinary<moore::LeOp>(lhs, rhs);
+    case BinaryOperator::LessThan:
+      return createBinary<moore::LtOp>(lhs, rhs);
+
+    // See IEEE 1800-2017 § 11.4.7 "Logical operators".
+    case BinaryOperator::LogicalAnd: {
+      // TODO: This should short-circuit. Put the RHS code into an scf.if.
+      lhs = convertToBool(lhs);
+      rhs = convertToBool(rhs);
+      if (!lhs || !rhs)
+        return {};
+      return builder.create<moore::AndOp>(loc, lhs, rhs);
+    }
+    case BinaryOperator::LogicalOr: {
+      // TODO: This should short-circuit. Put the RHS code into an scf.if.
+      lhs = convertToBool(lhs);
+      rhs = convertToBool(rhs);
+      if (!lhs || !rhs)
+        return {};
+      return builder.create<moore::OrOp>(loc, lhs, rhs);
+    }
+    case BinaryOperator::LogicalImplication: {
+      // `(lhs -> rhs)` equivalent to `(!lhs || rhs)`.
+      lhs = convertToBool(lhs);
+      rhs = convertToBool(rhs);
+      if (!lhs || !rhs)
+        return {};
+      auto notLHS = builder.create<moore::NotOp>(loc, lhs);
+      return builder.create<moore::OrOp>(loc, notLHS, rhs);
+    }
+    case BinaryOperator::LogicalEquivalence: {
+      // `(lhs <-> rhs)` equivalent to `(lhs && rhs) || (!lhs && !rhs)`.
+      lhs = convertToBool(lhs);
+      rhs = convertToBool(rhs);
+      if (!lhs || !rhs)
+        return {};
+      auto notLHS = builder.create<moore::NotOp>(loc, lhs);
+      auto notRHS = builder.create<moore::NotOp>(loc, rhs);
+      auto both = builder.create<moore::AndOp>(loc, lhs, rhs);
+      auto notBoth = builder.create<moore::AndOp>(loc, notLHS, notRHS);
+      return builder.create<moore::OrOp>(loc, both, notBoth);
+    }
```

Comment:
> Just wondering: what's the rationale of not having implication operations? Of course, there has to be some cut-off on what should be its own operation and which ones would be decomposed to more elementary ops. `Comb` goes the quite minimalistic route of not even having a `not` operation (which makes sense for its use-case). Where does/should `moore` position itself on this spectrum and why?

May I ask a question? Why does `Comb` not have a `not` operation?

Reply:
There is a short section about this in the rationale: https://circt.llvm.org/docs/Dialects/Comb/RationaleComb/#no-complement-negate-zext-sext-operators

Reply:
Thanks so much. I have to get familiar with these rationale.

Reply:
@maerhart This is a great point. At the moment it's a bit arbitrary, since most of these ops only work on integers. Arguably the reduction operators could also all just be more fundamental operators, but in that case you could argue in favor of them for IR compactness. The implication ops don't have that kind of explosion, they are truly just syntactic sugar.

Some of these operators are overloadable. I'm not entirely sure how they are represented in Slang's AST. For example, if you add two class instances that have the add operator overloaded (is that even a thing in SV?). Does that still show up as a regular "add" expression in the AST, or does Slang already resolve that to the operator overload.

I'd say create as few ops as possible, without causing any IR bloat like the reductions would. Very unprincipled :shrug:

Reply:
Op choices are an art informed by decades of experience from getting it wrong.  You want to model at the level of behavior you are interested in transforming.  You don't want redundancy.  You want compactness.  You don't want syntactic sugar.  You don't want to have to infer (lift) common cases from a sea of ops.  There are lots of conflicting goals.

An easy heuristic is that the number of ops to represent a concept should not scale with the size of the data type involved.  Hence reductions are fine, matching a sea of extracts of O(sizeof(type)) is bad.  add of constant is good, sequence of increments is bad.  xor of -1 is fine for not, since that isn't dependent on sizeof type.

Reply:
> There is a short section about this in the rationale: https://circt.llvm.org/docs/Dialects/Comb/RationaleComb/#no-complement-negate-zext-sext-operators

Thanks again for letting me know the benefit of rationales :smile: .